lib/autoproj/daemon/buildconf_manager.rb

```diff
@@ -0,0 +1,250 @@
+                    '-m',
+                    OVERRIDES_COMMIT_MSG,
+                    "refs/heads/#{branch_name}",
+                    commit_id
+                )
+                @importer.run_git_bare(
+                    @main, 'push', '-fu', @importer.remote_name, branch_name
+                )
+
+                created = client.branch(buildconf.owner, buildconf.name, branch_name)
+                cache.add(pull_request, overrides)
+                created
+            end
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Array<PackageRepository>]
+            def packages_affected_by_pull_request(pull_request)
+                packages.select do |pkg|
+                    pkg.name == pull_request.base_name &&
+                        pkg.owner == pull_request.base_owner &&
+                        pkg.branch == pull_request.base_branch
+                end
+            end
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Hash]
+            def overrides_for_pull_request(pull_request)
+                # TODO: Extract overrides from PR body.
+                #    This is not as easy as one could have thought
+                #    because the depended upon PR may have dependencies
+                #    of its own. We have to do it recursively. Detecting
+                #    changes in this chain of overrides is even more
+                #    challenging. Or, one could add the entire dependency
+                #    tree in every PR body.
+                packages_affected_by_pull_request(pull_request).map do |pkg|
+                    key = if pkg.package_set?
+                              "pkg_set:#{pkg.vcs[:repository_id]}"
+                          else
+                              pkg.package
+                          end
+                    {
+                        key => {
+                            'github' => "#{pull_request.head_owner}/"\
+                                        "#{pull_request.head_name}",
+                            'remote_branch' => pull_request.head_branch
+                        }
+                    }
+                end
+            end
+
+            # @param [Github::Branch] branch
+            # @return [Github::PullRequest, nil]
+            def pull_request_by_branch(branch)
+                pull_requests.find do |pr|
+                    branch_name_by_pull_request(pr) == branch.branch_name
+                end
+            end
+
+            # @param [Github::Branch] branch
+            # @return [nil]
+            def trigger_build(branch)
+                Autoproj.message "Triggering build for #{branch.branch_name} (new)"
+                nil
+            end
+
+            # @param [Array<Github::Branch>] branches
+            # @return [nil]
+            def trigger_build_if_branch_changed(branches)
+                branches.each do |branch|
+                    pr = pull_request_by_branch(branch)
```

Comment:
`@branches` has all branches from the buildconf. The daemon creates the branches with `autoproj/<owner>/<repo name>/pulls/<pr number>` so we can know the repository and the PR from the branch name without having to keep the branches cached as well

Reply:
This assumes that one PR == one branch, which is not true.

Reply:
It's not? My understanding was that we wanted to use branches on the buildconf to track each PR on the packages and that's what I implemented. Could you explain the functionality you had in mind again?

Reply:
You are right I believe ... Just didn't see it your way.

Because of the dependencies, I thought we would run a build for whole cliques of PRs. That is, a pull request that is "depended upon" won't have its own build - to reduce the amount of work.

Now, it's true that even under this scheme there is a "master" PR for every build ... so you're right ;-)

And, if you have for now setup the daemon to do a build per PR, let's finish this first and optimize later ...

Reply:
Ok, cool. :+1: